extensions/amp-carousel/0.1/slidescroll.js

```diff
@@ -57,9 +96,81 @@ export class AmpSlideScroll extends BaseCarousel {
 
   /** @override */
   hasPrev() {
+    return this.slideIndex_ > 0;
   }
 
   /** @override */
   hasNext() {
+    return this.slideIndex_ < this.slides_.length - 1;
+  }
+
+  /** @override */
+  goCallback(dir, unusedAnimate) {
+    if (this.slideIndex_ != null) {
+      if ((dir == DIR_FORWARD && this.hasNext()) ||
+          (dir == DIR_REVERSE && this.hasPrev())) {
+        this.mutateElement(() => {
+          this.showSlides_(this.slideIndex_ + dir);
+        }, this.slidesContainer_);
+      }
+    }
+  }
+
+  /**
+   * Makes the slide corresponding to the given index and the slides surrounding
+   *    it available for display.
+   * @param {number} newindex Index of the slide to be displayed.
+   * @private
+   */
+  showSlides_(newindex) {
+    const noOfSlides = this.noOfSlides_;
+    if (newindex < 0 ||
+        newindex >= this.noOfSlides_ ||
+        this.slideIndex_ == newindex) {
+      return;
+    }
+    const showIndexArr = [];
+    let newScrollLeft;
+    if (newindex == noOfSlides - 1) {
+      // Last slide.
+      showIndexArr.push(noOfSlides - 1, noOfSlides - 2);
+    } else if (newindex === 0) {
+      // First slide.
+      showIndexArr.push(0, 1);
+      newScrollLeft = 0;
+    } else {
+      showIndexArr.push(newindex - 1, newindex, newindex + 1);
+    }
+
+    if (newScrollLeft == null) {
+      newScrollLeft = this.slideWidth_;
```

Comment:
I'm confused by this. Wouldn't this always scroll to the very end of the slider (except when showing the first slide, since we set `newScrollLeft = 0`)?